File: packages/shared/session-chat-presentation/system-cards.ts
import { SESSION_CHAT_CODEX_GOAL_ID_PREFIX } from '@/packages/core-ui/chat/session-chat-pending';
import { SESSION_CHAT_FORK_BOUNDARY_ID_PREFIX, type SessionChatMessage } from '../session-chat';
import { agentDisplayName, parseSessionChatAgentMessage } from './agent-message';

/** The id `sessionChatPendingMessages` writes for a local command's captured output. */
const APP_COMMAND_OUTPUT_ID_PREFIX = 'app-command-output:';
/** The id it writes for a command Ghostex ran on the session's behalf, of which the rename is the one with its own card. */
const APP_COMMAND_ID_PREFIX = 'app-command:';
const AUTO_NAMED_TITLE_LEAD = 'Ghostex auto named this session';

/**
 * A system row that is a card or a rule rather than a sentence, with everything
 * the card shows. `marker` is the ordinary case: one muted line of the text the
 * daemon wrote.
 */
export type SessionChatSystemCard =
  | { kind: 'auto-named'; title: string }
  | { kind: 'fork-boundary'; text: string }
  | { kind: 'goal'; status: string; objective: string; usage: string }
  | { kind: 'command-output'; command: string; output: string }
  | { kind: 'agent-message'; sender: string; name: string; body: string }
  | { kind: 'marker'; text: string };

function blockText(message: SessionChatMessage, index: number): string {
  const block = message.blocks[index];
  return block?.type === 'text' ? block.text : '';
}

/**
 * Which card a system turn renders as, or null when the turn is not a system
 * turn. Both renderers classify here so a row cannot be a titled card in one
 * and raw prose in the other; each renderer owns only the layout of the card it
 * is handed.
 *
 * Suppressed harness turns are decided first by `sessionChatSuppressedTurnPresentation`
 * and never reach this.
 */
export function classifySessionChatSystemCard(
  message: SessionChatMessage,
  markdown: string
): SessionChatSystemCard | null {
  if (message.role !== 'system') {
    return null;
  }
  const autoNamedTitle =
    message.id.startsWith(APP_COMMAND_ID_PREFIX) && blockText(message, 0) === AUTO_NAMED_TITLE_LEAD
      ? blockText(message, 1).trim()
      : '';
  if (autoNamedTitle.length > 0) {
    return { kind: 'auto-named', title: autoNamedTitle };
  }
  if (message.id.startsWith(SESSION_CHAT_FORK_BOUNDARY_ID_PREFIX)) {
    return { kind: 'fork-boundary', text: markdown };
  }
  if (message.id.startsWith(SESSION_CHAT_CODEX_GOAL_ID_PREFIX)) {
    return {
      kind: 'goal',
      status: blockText(message, 0),
      objective: blockText(message, 1),
      usage: blockText(message, 2),
    };
  }
  if (message.id.startsWith(APP_COMMAND_OUTPUT_ID_PREFIX)) {
    return { kind: 'command-output', command: blockText(message, 0), output: blockText(message, 1) };
  }
  const agentMessage = parseSessionChatAgentMessage(markdown);
  if (agentMessage) {
    return {
      kind: 'agent-message',
      sender: agentMessage.sender,
      name: agentDisplayName(agentMessage.sender),
      body: agentMessage.body,
    };
  }
  return { kind: 'marker', text: markdown };
}

File: packages/gx-chat-core/src/transcript/system_cards.rs
//! Which card a system turn renders as.
//!
//! Ported from `packages/shared/session-chat-presentation/system-cards.ts`. Both renderers classify
//! here so a row cannot be a titled card in one and raw prose in the other; each renderer owns only
//! the layout of the card it is handed.

use ghostex_gx_protocol::{ChatBlock, ChatMessage, ChatRole};
use serde_json::{json, Value};

use crate::transcript::agent_message::{agent_display_name, parse_agent_message};
use crate::transcript::foreign::{CODEX_GOAL_ID_PREFIX, FORK_BOUNDARY_ID_PREFIX};
use crate::transcript::jsstr::js_trim;

/// The id `sessionChatPendingMessages` writes for a local command's captured output.
const APP_COMMAND_OUTPUT_ID_PREFIX: &str = "app-command-output:";
/// The id it writes for a command Ghostex ran on the session's behalf, of which the rename is the
/// one with its own card.
const APP_COMMAND_ID_PREFIX: &str = "app-command:";
const AUTO_NAMED_TITLE_LEAD: &str = "Ghostex auto named this session";

fn block_text(message: &ChatMessage, index: usize) -> &str {
    match message.blocks.get(index) {
        Some(ChatBlock::Text { text }) => text,
        _ => "",
    }
}

/// Which card a system turn renders as, or `Value::Null` when the turn is not a system turn.
///
/// Suppressed harness turns are decided first by the noise classifier and never reach this.
/// `marker` is the ordinary case: one muted line of the text the daemon wrote.
pub fn classify_system_card(message: &ChatMessage, markdown: &str) -> Value {
    if message.role != ChatRole::System {
        return Value::Null;
    }
    let auto_named_title = if message.id.starts_with(APP_COMMAND_ID_PREFIX)
        && block_text(message, 0) == AUTO_NAMED_TITLE_LEAD
    {
        js_trim(block_text(message, 1))
    } else {
        ""
    };
    if !auto_named_title.is_empty() {
        return json!({ "kind": "auto-named", "title": auto_named_title });
    }
    if message.id.starts_with(FORK_BOUNDARY_ID_PREFIX) {
        return json!({ "kind": "fork-boundary", "text": markdown });
    }
    if message.id.starts_with(CODEX_GOAL_ID_PREFIX) {
        return json!({
            "kind": "goal",
            "status": block_text(message, 0),
            "objective": block_text(message, 1),
            "usage": block_text(message, 2),
        });
    }
    if message.id.starts_with(APP_COMMAND_OUTPUT_ID_PREFIX) {
        return json!({
            "kind": "command-output",
            "command": block_text(message, 0),
            "output": block_text(message, 1),
        });
    }
    if let Some(agent_message) = parse_agent_message(markdown) {
        return json!({
            "kind": "agent-message",
            "sender": agent_message.sender,
            "name": agent_display_name(&agent_message.sender),
            "body": agent_message.body,
        });
    }
    json!({ "kind": "marker", "text": markdown })
}
